cpp/src/utils/metrics/metric.hh

```diff
@@ -0,0 +1,50 @@
+#ifndef TS_CPP_UTILS_METRICS_METRIC_HH_
+#define TS_CPP_UTILS_METRICS_METRIC_HH_
+
+#include <string>
+#include <vector>
+
+#include "src/utils/metrics/dimension.hh"
+
+namespace torchserve {
+    enum MetricType {
+        COUNTER,
+        GAUGE,
+        HISTOGRAM
+    };
+
+    class MetricValue {
+        public:
+        MetricValue(const double& value, const uint64_t& timestamp, const std::string& request_id);
+        double GetValue() const;
+        uint64_t GetTimestamp() const;
+        const std::string GetRequestId() const;
+
+        private:
+        const double value;
+        const uint64_t timestamp;
+        const std::string request_id;
+    };
+
+    class Metric {
```

Comment:
As we sync up last week, we need first define torchserve Metric API so that we are able to  extend it to different implementation such as frontend prometheus and backend log metrics.

Here the code is more related to backend. it is difficult to later extend to frontend. So please define metric api first, and then implement backend metrics.

Reply:
I believe we are on the same page. The class I'm defining here is going to be used to store the metric attributes `(name, unit, type, dimensions and values)` for a single instance of a particular metric. The plan is to add a `MetricCache` class which will hold several of these `Metric` objects. The modularity to emit metrics either as a log or to prometheus or a different metric framework in future will come from the `MetricEmitter` handle in the `MetricCache` instance. For example, the backend `MetricCache` instance will use the `STDOUTMetricEmitter` and the frontend `MetricCache` instance will use the `PrometheusMetricEmitter`. Please let me know let me know your thoughts on this.

Reply:
sounds good.

Reply:
Updated implementation to follow the API defined in metrics redesign RFC: https://github.com/pytorch/serve/issues/1492